doc/man/psql.1
.\" This is -*-nroff-*-
.\" XXX standard disclaimer belongs here....
.\" $Header: /cvsroot/pgsql/doc/man/Attic/psql.1,v 1.1.1.1 1996/08/18 22:14:26 scrappy Exp $
.TH PSQL UNIX 1/20/96 Postgres95 Postgres95
.SH NAME
psql \(em run the interactive query front-end
.SH SYNOPSIS
.BR psql
[\c
.BR "-a"
authsvc
]
[\c
.BR "-A"
]
[\c
.BR "-c"
query
]
[\c
.BR "-d"
dbName]
[\c
.BR "-e"
]
[\c
.BR "-f"
filename]
[\c
.BR "-h"
hostname]
[\c
.BR "-H"
]
[\c
.BR "-l"
port]
[\c
.BR "-n"
]
[\c
.BR "-o"
filename
]
[\c
.BR "-p"
port]
[\c
.BR "-q"
]
[\c
.BR "-s"
]
[\c
.BR "-S"
]
[\c
.BR "-t"
]
[\c
.BR "-x"
]
[dbname]
.in -5n
.SH DESCRIPTION
psql is a interactive query front-end to Postgres.  It enables you to
type in queries interactively, issue them to Postgres, and see the query
results.
.IR psql
can be used in a pipe sequence, and automatically detects when it
is not listening or talking to a real tty.
.IR psql
is designed to be an enhanced version of the older
.IR "monitor"
program.
.PP
.IR "psql"
is a frontend application, like any other.  Hence, a
.IR "postmaster"
process must be running on the database server host before
.IR "psql"
is executed.  In addition, the correct
.IR "postmaster"
port number must be specified
as described below.
.PP
The optional argument
.IR dbname
specifies the name of the database to be accessed.  This database must
already have been created.
.IR dbname
defaults to the value of the
.SM USER
environment variable or, if that's not set, to the Unix account name of the
current user.
.PP
.IR "psql"
understands the following command-line options:
.TP
.BR "-a" " system"
Specifies an authentication system
.IR "system"
(see
.IR introduction (1))
to use in connecting to the
.IR postmaster
process.  The default is site-specific.
.TP
.BR "-A"
Turn off fill justification when printing out attributes.
.TP
.BR "-c" " query"
Specifies that
.IR "psql"
is to execute one query string,
.IR "query" ,
and then exit.  This is useful for shell scripts, typically in
conjunction with the
.BR -q ""
options.
.BR -c
option in shell scripts.
.TP
.BR "-d" " dbName"
Specifies the name of the database to connect to.
.TP
.BR "-e" " "
Echo the query sent to the backend
.TP
.BR "-f" " filename"
Use the file
.IR "filename"
as the source of queries instead of reading queries interactively.
.TP
.BR "-h" " hostname"
Specifies the hostname of the machine on which the
.IR postmaster
is running.  Defaults to the name of the local host, or the value of
the
.SM PGHOST
environment variable (if set).
.TP
.BR "-H"
Turns on
.SM HTML3.0
tabular output.
.TP
.BR "-l"
Lists all available databases
.TP
.BR "-n"
Do not use the readline library for input line editing and command history.
.TP
.BR "-p" " port"
Specifies the Internet TCP port on which the
.IR postmaster
is listening for connections.  Defaults to 5432, or the value of the
.SM PGPORT
environment variable (if set).
.TP
.BR "-q"
Specifies that
.IR psql
should do its work quietly.  By default, it
prints welcome and exit messages and prompts for each query, and prints
out the number of rows returned from a query.
If this option is used, none of this happens. This is useful with the
.BR -c
option in shell scripts.
.TP
.BR "-s"
Run in single-step mode where the user at prompted for each query before
it is sent to the backend.
.TP
.BR "-S"
Run ins single-line mode where each query is terminated by a newline,
instead of a semicolon.
.TP
.BR "-t"
Turn off printing of attributes names.
This is useful with the
.BR -c
option in shell scripts.
.TP
.BR "-x"
Turns on extended field mode. When enabled each tuple will have its field
names printed on the left with the field values printed on the right.
This is useful for tuples which are otherwise too long to fit into
one screen line. HTML tuple output supports this mode also.
.PP
You may set environment variables to avoid typing some of the above
options.  See the
.SM "ENVIRONMENT VARIABLES"
section below.
.SH "CONNECTING TO A DATABASE"
.IR psql
attempts to make a connection to the database at the hostname and
port number specified on the command line.   If the connection could not
be made for any reason (e.g. insufficient privileges, postmaster is not
running on the server, etc)
.IR psql
will return an error that says
.nf
Connection to database failed.
.fi
The reason for the connection failure is not provided.
.SH "ENTERING QUERIES"
In normal operation, psql provides a prompt with the name of the
database that psql is current connected to followed by the string "=>".
For example,
.nf
Welcome to the POSTGRES95 interactive sql monitor:
  Please read the file COPYRIGHT for copyright terms of POSTGRES95

   type \e? for help on slash commands
   type \eq to quit
   type \eg or terminate with semicolon to execute query
 You are currently connected to the database: testdb

testdb=>
.fi
.PP
At the prompt, the user may type in SQL queries.  Unless the -S option
is set, input lines are sent to the backend when a query-terminating
semicolon is reached.
.PP
Whenever a query is executed, psql also polls for asynchronous notification
events generated by
.IR listen (l)
and
.IR notify (l).
.PP
.SH "PSQL COMMANDS"
.IP "\ea"
Toggle field alignment when printing out attributes.
.IP "\eC \fIcaption\fR"
Set the HTML3.0 table caption.
.IP "\ec \fIdbname\fR"
Establish a connection to a new database. The previous connection is closed.
.IP "\ed [\fItable\fR]"
List tables in the database, or if
.IR table
is specified, list the columns in
.IR table.
If table name is
.IR *,
list all tables and column information for each tables.
.IP "\ee [\fIfilename\fR]"
Edit the current query buffer or \fIfile\fR.
.IP "\eE [\fIfilename\fR]"
Edit the current query buffer or \fIfile\fR and execute it
upon editor exit.
.IP "\ef [\fIseparator\fR]"
Set the field separator.  Default is a single blank space.
.IP "\eg [\fI|command\fR] | [\fIfilename\fR]"
Send the current query input buffer to the backend and optionally
save the output in
.IR filename
or pipe the output into
.IR "|command".
.IP "\eh [\fIcommand\fR]"
Give syntax help on the specified SQL command.  If the
.IR command
is not specified, list all the commands for which syntax help is
available.  If the
.IR command
is
.IR *,
give syntax help on all SQL commands.
.IP "\eH"
Toggle html3 output.
.IP "\ei \fIfilename\fR"
Read queries from
.IR filename
into the query input buffer.
.IP "\el"
List all the databases in the server.
.IP "\em"
Toggle monitor-like table display.
This is standard SQL output (i.e extra border characters).
.IP "\eo [\fI|command\fR] | [\fIfilename\fR]"
Send query results to
.IR filename .
Or pipe into
.IR command .
If no arguments are specified, send query results to
.IR stdout .
.IP "\ep"
Print the current query buffer.
.IP \eq
Quit the psql program.
.IP "\er"
Reset(clear) the query buffer.
.IP "\es [\fIfilename\fR]"
Print or save the command line history to \fIfilename\fR.  (Only available if psql is
configured to use readline)
.IP "\et"
Toggle display of output attribute name headings and row count (defaults to on).
.IP "\eT"
Set html3.0 <table ...> options.
.IP "\ex"
Toggles extended field mode. When enabled each tuple will have its field
names printed on the left with the field values printed on the right.
This is useful for tuples which are otherwise too long to fit into
one screen line. HTML tuple output mode supports this flag too.
.IP "\e! [\fIcommand\fR]"
Escape to shell or execute
.IR command.
.IP \e?
Get help information about the \e commands.
.SH "ENVIRONMENT VARIABLES"
You may set any of the following environment variables to avoid
specifying command-line options:
.nf
hostname:	PGHOST
port:		PGPORT
tty:		PGTTY
options:		PGOPTION
realm:		PGREALM
.fi
.PP
If
.SM PGOPTION
is specified, then the options it contains are parsed
.BR before
any command-line options.
.PP
.SM PGREALM
only applies if
.IR Kerberos
authentication is in use.  If this environment variable is set, Postgres
will attempt authentication with servers for this realm and use
separate ticket files to avoid conflicts with local ticket files.  See
.IR introduction (1)
for additional information on
.IR Kerberos .
.PP
See
.IR introduction (libpq)
for additional details.
.SH "RETURN VALUE"
When executed with the
.BR "-c"
option,
.IR psql
returns 0 to the shell on successful query completion, 1 otherwise.
.IR psql
will also return 1 if the connection to a database could not be made for
any reason.
.SH "SEE ALSO"
introduction(libpq),
monitor(1)
postgres(1),
postmaster(1).
.SH BUGS
If multiple queries are sent to the backend at once without semicolon
termination after each query, psql gets confused about the query
results.  The queries will still be processed correctly by the backend.



HISTORY
                                 Release Notes

                                  Release 7.3

     Release date: 2002-11-??

Overview

   Major changes in this release:

   Schemas

           Schemas allow users to create objects in their own namespace
           so two people or applications can have tables with the same
           name. There is also a public schema for shared tables.
           Table/index creation can be restricted by removing
           permissions on the public schema.

   Drop Column

           PostgreSQL now supports ALTER TABLE ... DROP COLUMN functionality.

   Table Functions

           Functions returning multiple rows and/or multiple columns are
           now much easier to use than before.  You can call such a
           "table function" in the SELECT FROM clause, treating its output
           like a table. Also, plpgsql functions can now return sets.

   Prepared Queries

           For performance, PostgreSQL now supports prepared queries.

   Dependency Tracking

           PostgreSQL now records object dependencies, which allows
           improvements in many areas.

   Privileges

           Functions and procedural languages now have privileges, and
           people running them can take on the privileges of their creators.

   Multibyte/Locale

           Both multibyte and locale are now always enabled.

   Logging

           A variety of logging options have been enhanced.

   Interfaces

           A large number of interfaces have been moved to 
           http://gborg.postgresql.org where they can be developed
           and released independently.

   Functions/Identifiers

           By default, functions can now take up to 32 parameters, and 
           identifiers can be up to 63 bytes long.

     ----------------------------------------------------------------------

Migration to version 7.3

   A dump/restore using "pg_dump" is required for those wishing to migrate
   data from any previous release.  If your application examines the
   system catalogs, additional changes will be required due to the
   introduction of schemas in 7.3; for more information, see:

		http://www.ca.postgresql.org/docs/momjian/upgrade_tips_7.3

   Observe the following incompatibilities:

     * Pre-6.3 clients are no longer supported.

     * pg_hba.conf now has a USER column and additional features.

     * Several GUC logging parameters have been renamed and improved.

     * LIMIT #,# is disabled, use LIMIT # OFFSET #.

     * INSERTs with column lists must specify all values, e.g. 
       INSERT INTO tab (col1, col2) VALUES ('val1') is now invalid.

     * DROP object now takes either CASCADE or RESTRICT to control
       whether dependent objects are also dropped.

     * An index is now not automatically created for SERIAL columns.

     * A SET inside an aborted transaction is now rolled back.

     * COPY no longer considers missing trailing columns to be NULL.

     * TIMESTAMP and TIME data types now default to WITHOUT TIMEZONE.

     * Pre-7.3 databases loaded into 7.3 will not have the new object
       dependencies for SERIAL, UNIQUE constraints, and foreign keys. See
       /contrib/adddepend for a detailed description and a script that
       will add the such dependencies.

     ----------------------------------------------------------------------

Changes

2002-10-23

Server Operation
================
Kerberos V Heimdal support (KTH)
Add pg_locks view to show locks (Neil)
Security fixes for password negotiation memory allocation (Neil)
Remove support for version 0, <=6.2 FE/BE protocol (Tom)
Reserve the last few backend slots for superusers, add GUC variable
  superuser_reserved_connections to control this (Nigel J. Andrews)

Performance
===========
Improve startup by calling localtime() only once (Tom)
Cache system catalog information in flat files for faster startup (Tom)
Improve caching of index information (Tom)
Optimizer improvements (Tom, Fernando Nasser)
Catalog caches now store failed lookups (Tom)
Hash function improvements (Neil)
Improve performance of query tokenization and network handling (Peter)
Speed improvement for large object restore (Mario Weilguni)
Mark expired index entries on first lookup, saving later heap fetches (Tom)
Eliminate NULL bitmap padding when not required (Manfred)
Add BSD-licensed qsort() for Solaris, for performance (Bruce)
Reduce per-row overhead by four bytes (Manfred Koizar)
Fix GEQO optimizer bug (Neil Conway)
Make WITHOUT OID actually save four bytes per row (Manfred Koizar)
Add GUC default_statistics_target variable to specify ANALYZE buckets (Neil)
Use local buffer cache for temporary tables so no WAL overhead (Tom)
Improve free space map performance on large tables (Stephen Marshall, Tom)
Improved WAL write concurrency (Tom)

Privileges
==========
Add privileges on functions and procedural languages (Peter)
Add OWNER to CREATE DATABASE so superusers can create databases
  on behalf of unprivileged users.  (Gavin Sherry, Tom)
Add new object permission bits EXECUTE and USAGE (Tom)
Add SET SESSION AUTHORIZATION DEFAULT and RESET SESSION AUTHORIZATION (Peter)
Allow functions to be executed with the privilege of the function owner (Peter)

Client Authentication
=====================
Allow $HOME/.pgpass to store host/user password combinations (Alvaro Herrera)

Server Configuration
====================
Server log messages now tagged with LOG, not DEBUG (Bruce)
Add USER column to pg_hba.conf (Bruce)
Have log_connections output two lines in log file (Tom)
Remove debug_level from postgresql.conf, now SERVER_MIN_MESSAGES (Bruce)
New ALTER DATABASE/USER ... SET command for per-user/db initialization (Peter)
New GUC variables server_min_messages and client_min_messages which control
  which messages are sent to the server logs or client applications (Bruce)
Allow pg_hba.conf to specify lists of users/dbs separated by commas, 
  group names prepended with +, and file names prepended with @ (Bruce)

src/interfaces/jdbc/example/psql.java
package example;

import java.io.*;
import java.sql.*;
import java.text.*;

/*
 * This example application demonstrates some of the drivers other features
 * by implementing a simple psql replacement in Java.
 *
 */

public class psql
{
	Connection	db;		// The connection to the database
	Statement	st;		// Our statement to run queries with
	DatabaseMetaData dbmd;	// This defines the structure of the database
	boolean done = false;		  // Added by CWJ to permit \q command

	public psql(String args[]) throws ClassNotFoundException, FileNotFoundException, IOException, SQLException
	{
		String url = args[0];
		String usr = args[1];
		String pwd = args[2];

		// Load the driver
		Class.forName("org.postgresql.Driver");

		// Connect to database
		System.out.println("Connecting to Database URL = " + url);
		db = DriverManager.getConnection(url, usr, pwd);

		dbmd = db.getMetaData();
		st = db.createStatement();

		// This prints the backend's version
		System.out.println("Connected to " + dbmd.getDatabaseProductName() + " " + dbmd.getDatabaseProductVersion());

		System.out.println();

		// This provides us the means of reading from stdin
		StreamTokenizer input = new StreamTokenizer(new InputStreamReader(System.in));
		input.resetSyntax();
		input.slashSlashComments(true); // allow // as a comment delimiter
		input.eolIsSignificant(false);	// treat eol's as spaces
		input.wordChars(32, 126);
		input.whitespaceChars(59, 59);
		// input.quoteChar(39); *** CWJ: messes up literals in query string ***

		// Now the main loop.
		int tt = 0, lineno = 1;
		while (tt != StreamTokenizer.TT_EOF && ! done)
		{ // done added by CWJ to permit \q command
			System.out.print("[" + lineno + "] ");
			System.out.flush();

			// Here, we trap SQLException so they don't terminate the application
			try
			{
				if ((tt = input.nextToken()) == StreamTokenizer.TT_WORD)
				{
					processLine(input.sval);
					lineno++;
				}
			}
			catch (SQLException ex)
			{
				System.out.println(ex.getMessage());
			}
		}

		System.out.println("Now closing the connection");
		st.close();
		db.close();

	}

	/*
	 * This processes a statement
	 */
	public void processLine(String line) throws SQLException
	{
		if (line.startsWith("\\"))
		{
			processSlashCommand(line);
			return ;
		}

		boolean type = st.execute(line);
		boolean loop = true;
		while (loop)
		{
			if (type)
			{
				// A ResultSet was returned
				ResultSet rs = st.getResultSet();
				displayResult(rs);
			}
			else
			{
				int count = st.getUpdateCount();

				if (count == -1)
				{
					// This indicates nothing left
					loop = false;
				}
				else
				{
					// An update count was returned
					System.out.println("Updated " + st.getUpdateCount() + " rows");
				}
			}

			if (loop)
				type = st.getMoreResults();
		}
	}

	/*
	 * This displays a result set.
	 * Note: it closes the result once complete.
	 */
	public void displayResult(ResultSet rs) throws SQLException
	{
		ResultSetMetaData rsmd = rs.getMetaData();

		// Print the result column names
		int cols = rsmd.getColumnCount();
		for (int i = 1;i <= cols;i++)
			System.out.print(rsmd.getColumnLabel(i) + (i < cols ? "\t" : "\n"));

		// now the results
		while (rs.next())
		{
			for (int i = 1;i <= cols;i++)
			{
				Object o = rs.getObject(i);
				if (rs.wasNull())
					System.out.print("{null}" + (i < cols ? "\t" : "\n"));
				else
					System.out.print(o.toString() + (i < cols ? "\t" : "\n"));
			}
		}

		// finally close the result set
		rs.close();
	}

	/*
	 * This process / commands (for now just /d)
	 */
	public void processSlashCommand(String line) throws SQLException
	{
		if (line.startsWith("\\d"))
		{

			if (line.startsWith("\\d "))
			{
				// Display details about a table
				String table = line.substring(3);
				displayResult(dbmd.getColumns(null, null, table, "%"));
			}
			else
			{
				String types[] = null;
				if (line.equals("\\d"))
					types = allUserTables;
				else if (line.equals("\\di"))
					types = usrIndices;
				else if (line.equals("\\dt"))
					types = usrTables;
				else if (line.equals("\\ds"))
					types = usrSequences;
				else if (line.equals("\\dS"))
					types = sysTables;
				else
					throw new SQLException("Unsupported \\d command: " + line);

				// Display details about all system tables
				//
				// Note: the first two arguments are ignored. To keep to the spec,
				//		 you must put null here
				//
				displayResult(dbmd.getTables(null, null, "%", types));
			}
		}
		else if (line.equals("\\q")) // Added by CWJ to permit \q command
			done = true;
		else
			throw new SQLException("Unsupported \\ command: " + line);
	}

	private static final String allUserTables[] = {"TABLE", "INDEX", "SEQUENCE"};
	private static final String usrIndices[] = {"INDEX"};
	private static final String usrTables[] = {"TABLE"};
	private static final String usrSequences[] = {"SEQUENCE"};
	private static final String sysTables[] = {"SYSTEM TABLE", "SYSTEM INDEX"};

	/*
	 * Display some instructions on how to run the example
	 */
	public static void instructions()
	{
		System.out.println("\nThis example shows how some of the other JDBC features work within the\ndriver. It does this by implementing a very simple psql equivalent in java.\nNot everything that psql does is implemented.\n");
		System.out.println("Useage:\n java example.psql jdbc:postgresql:database user password [debug]\n\nThe debug field can be anything. It's presence will enable DriverManager's\ndebug trace. Unless you want to see screens of items, don't put anything in\nhere.");
		System.exit(1);
	}

	/*
	 * This little lot starts the test
	 */
	public static void main(String args[])
	{
		System.out.println("PostgreSQL psql example v6.3 rev 1\n");

		if (args.length < 3)
			instructions();

		// This line outputs debug information to stderr. To enable this, simply
		// add an extra parameter to the command line
		if (args.length > 3)
			DriverManager.setLogStream(System.err);

		// Now run the tests
		try
		{
			psql test = new psql(args);
		}
		catch (Exception ex)
		{
			System.err.println("Exception caught.\n" + ex);
			ex.printStackTrace();
		}
	}
}
